units/provisioning_setup_host.go

```diff
@@ -898,3 +916,111 @@ func (j *setupHostJob) tryRequeue(ctx context.Context) {
 }
+
+func attachVolume(ctx context.Context, env evergreen.Environment, h *host.Host) error {
+	if !(h.Distro.JasperCommunication() && h.Distro.IsLinux()) {
+		return errors.Errorf("host '%s' of distro '%s' doesn't support mounting a volume", h.Id, h.Distro.Id)
+	}
+
+	if h.HomeVolumeID == "" {
+		mgrOpts, err := cloud.GetManagerOptions(h.Distro)
+		if err != nil {
+			return errors.Wrapf(err, "can't get ManagerOpts for '%s'", h.Id)
+		}
+		cloudMgr, err := cloud.GetManager(ctx, env, mgrOpts)
+		if err != nil {
+			return errors.Wrapf(err,
+				"failed to get cloud manager for host %s with provider %s",
+				h.Id, h.Provider)
+		}
+
+		// create the volume
+		volume, err := cloudMgr.CreateVolume(ctx, &host.Volume{
+			Size:             h.HomeVolumeGB,
+			AvailabilityZone: h.Zone,
+			CreatedBy:        h.StartedBy,
+			Type:             evergreen.DefaultEBSType,
+		})
+		if err != nil {
+			return errors.Wrapf(err, "can't create a new volume for host '%s'", h.Id)
+		}
+
+		// attach to the host
+		attachment := host.VolumeAttachment{VolumeID: volume.ID}
+		if err = cloudMgr.AttachVolume(ctx, h, &attachment); err != nil {
+			return errors.Wrapf(err, "can't attach volume '%s' to host '%s'", volume.ID, h.Id)
+		}
+
+		if err = h.SetHomeVolume(volume.ID, attachment.DeviceName); err != nil {
+			return errors.Wrapf(err, "can't set home volume for host '%s'", h.Id)
+		}
+	}
+
+	// run the distro's mount script
+	var err error
+	for i := 0; i < mountRetryLimit; i++ {
+		if err = errors.Wrap(mountLinuxVolume(ctx, env, h), "can't mount volume"); err == nil {
+			return nil
+		}
+		time.Sleep(mountSleepDuration)
+	}
+	return err
+}
+
+func mountLinuxVolume(ctx context.Context, env evergreen.Environment, h *host.Host) error {
+	client, err := h.JasperClient(ctx, env)
+	if err != nil {
+		return errors.Wrap(err, "can't get Jasper client")
+	}
+	defer func() {
+		grip.Warning(message.WrapError(client.CloseConnection(), message.Fields{
+			"message": "could not close connection to Jasper",
+			"host":    h.Id,
+			"distro":  h.Distro.Id,
+		}))
+	}()
+
+	deviceName := h.HomeVolumeDeviceName
+	if h.Distro.HomeVolumeDeviceName != "" {
+		deviceName = h.Distro.HomeVolumeDeviceName
+	}
+	proc, err := client.CreateProcess(ctx, &options.Create{Args: []string{"sudo", "umount", "-f", fmt.Sprintf("/dev/%s", deviceName)}})
+	if err != nil {
+		return errors.Wrap(err, "problem creating process for umount")
+	}
+	// continue on umount mount error
+	exitCode, err := proc.Wait(ctx)
+	if err != nil && exitCode != umountMountErrorCode {
+		return errors.Wrap(err, "problem waiting for umount command")
+	}
+
+	commands := [][]string{[]string{"sudo", "/sbin/mkfs.xfs", "-f", fmt.Sprintf("/dev/%s", deviceName)}}
+	commands = append(commands, []string{"sudo", "mkdir", "-p", fmt.Sprintf("/%s", evergreen.HomeVolumeDir)})
+	commands = append(commands, []string{"sudo", "mount", fmt.Sprintf("/dev/%s", deviceName), fmt.Sprintf("/%s", evergreen.HomeVolumeDir)})
```

Comment:
One last question - are these commands all idempotent? For example, if mkfs.xfs succeeds but a later command fails, can mkfs.xfs still be run on retries? I wouldn't want these to run and a command fails later on, the script should retry but not fail because an operation was already completed. I feel like if they are not idempotent, each of these commands should be retried one by one until they succeed (at which point they can move on to the next command) or hit the retry limit.

Reply:
I intended for attachVolume to be idempotent: 
- we only create and attach a volume to the host if we haven't yet (so we don't have a new volume for every host provisioning retry in the non-userdata case).
- we unmount the partition so if mounting succeeded in a previous retry we can still format the filesystem and mount will succeed.
- `mkdir -p` exits 0 even if the directory already exists